### PR TITLE
Fix nrf52840 serial mentioned #1021

### PR DESCRIPTION
**Pull request by hathach**
- update tinyusb for wanted char
- move usb code into usb.c

### CODE REVIEW

**Comment by hathach** [comment]
sorry there is a bit off issue, the current code will swallow the ctrl+c as well, hold on, I will fix it.

**Comment by hathach** [comment]
should be OK now. I am not sure if the `lib/utils/interrupt_char.c` is OK to edit. So I add a separated interrupt_char.c with modification under nrf/ . It only need to tell the usb stack to trigger callback when receiving the Ctrl+C

**Comment by hathach** [comment]
#1021

File: ports/nrf/usb/usb.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef MICROPY_INCLUDED_NRF_USB_H
+#define MICROPY_INCLUDED_NRF_USB_H
+
+#ifdef NRF52840_XXAA
```
Comment (arturo182) [inline ports/nrf/usb/usb.h]: Please remove

File: ports/nrf/usb/usb.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ * THE SOFTWARE.
+ */
+
+#include "nrfx.h"
+#include "nrfx_power.h"
+#include "tick.h"
+#include "usb.h"
+#include "lib/utils/interrupt_char.h"
+
+#ifdef NRF52840_XXAA
```
Comment (arturo182) [inline ports/nrf/usb/usb.c]: Please remove

File: ports/nrf/usb/usb.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+{
+    .device     = NULL,
+    .config     = NULL,
+    .string_arr = (uint8_t const **) string_desc_arr,
+    .hid_report = NULL
+};
+
+void usb_init(void) {
+
+#ifdef SOFTDEVICE_PRESENT
```
Comment (arturo182) [inline ports/nrf/usb/usb.c]: I don't think this ifdef is needed, I did the fix for having usb and SD co-exist and no extra code is needed here.

Comment (hathach) [reply ports/nrf/usb/usb.c]: the part when SD is enabled is currently handled inside the stack. 
https://github.com/hathach/tinyusb/blob/develop/src/portable/nordic/nrf5x/hal_nrf5x.c#L178

Previously they are all inside in tusb hal, then recent Nordic sdk break non-SD api by introducing nrfx (SD API is more stable). But I think I will move it out of the stack, making the stack more generic. I will take this chance to do this

Comment (hathach) [reply ports/nrf/usb/usb.c]: Plot all is planned here 
 https://github.com/hathach/tinyusb/blob/develop/src/portable/nordic/nrf5x/hal_nrf5x.c#L180

**Comment by hathach** [comment]
@arturo182 do we define SOFTDEVICE_PRESENT when  SD is selected, I couldn't find it within the makefile

**Comment by arturo182** [comment]
It actually looks like we don't define it, it's defined usually in the Nordic SDK which we don't use. Should probably define it in the Makefile if SD is not empty, because even nRFx checks for it sometimes.

**Comment by hathach** [comment]
@arturo182 it uses throughout Nordic code, it is good idea to use it to detect SD existence as well. I think for peripheral that is shared with SD, nrfx will need that to detect the SD existed and enabled.

**Comment by hathach** [comment]
btw, could you brief me on how to run a simple ble script e.g advertising something. I will check if changes break anything when the SD is running.

**Comment by arturo182** [comment]
For sure!
The simplest way I can think of to check if BLE works is:

```
import ubluepy
p = ubluepy.Peripheral()
p.advertise(device_name='Python')
```

After that you should be able to see a device named `Python` when scanning, I use https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=en to test it.

**Comment by arturo182** [comment]
@tannewt I can approve but not merge:
> Only those with write access to this repository can merge pull requests.

**Comment by bobbahatti** [comment]
https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp

**Pull request merged**